### DISCUSSION TITLE
MQTT disconnects and rule events

### DISCUSSION THREAD

**Discussion by sciurius** [arendst/Tasmota#17278] [Q&A]
In an attempt to track down unsollicited MQTT disconnects, I added the following rule:

````
ON Mqtt#Disconnected DO Var3 %timestamp% ENDON
ON Mqtt#Connected DO Publish tele/%topic%/BLACKOUT {"From":"%Var3%","To":"%timestamp%"} ENDON
````
I receive several messages similar to
````
{"From":"2022-12-04T20:08:39","To":"2022-12-04T20:08:49"}
{"From":"2022-12-04T20:08:39","To":"2022-12-04T21:21:27"}
{"From":"2022-12-04T20:08:39","To":"2022-12-04T21:23:29"}
{"From":"2022-12-04T20:08:39","To":"2022-12-04T21:24:31"}
````
Since there are no signs of Wifi dis/reconnects, this seems to indicate either MQTT connects without disconnect, or failures in raising the `mqtt#disconnect` event and/or handling the rule.

Is this a bug, or do I misunderstand how it is supposed to work?

**Comment by sfromis** [comment]
For context, you "forgot" to mention the circumstance that this is related to reconnecting to MQTT, without clear signs of the Wifi being out.

**Reply by sciurius** [reply]
Thanks. I've updated the original question.

**Reply by barbudor** [reply]
If I get your point, you are assuming that there was no mqtt#disconnected events because the Var3 has the same values for all 3 events
IT would be interresting to have logs for that moment. 
If you are sure that Wifi is not in cause, you can use the web console, or a syslog
If wifi is possibly in cause, then serial would be the only way

In case of disconnection, there should be a message like `MQTT connect failed to <host>:<port>, rc <error>`

**Comment by sciurius** [comment]
This is a typical series of console messages:
````
18:59:05.547 MQT: tele/tasm07/STATE = {"Time":"2022-12-05T18:59:05","Uptime":"1T20:25:18","UptimeSec":159918,"Heap":21,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":30,"Wifi":{"AP":1,"SSId":"Oh Denneboom","BSSId":"******","Channel":1,"Mode":"11n","RSSI":90,"Signal":-55,"LinkCount":2,"Downtime":"0T00:00:05"}}
18:59:05.557 MQT: tele/tasm07/SENSOR = {"Time":"2022-12-05T18:59:05","AM2301":{"Temperature":22.0,"Humidity":36.2,"DewPoint":6.3},"TempUnit":"C"}
18:59:44.509 MQT: Attempting connection...
18:59:44.639 MQT: Connected
18:59:44.643 MQT: tele/tasm07/LWT = Online (retained)
18:59:44.646 MQT: cmnd/tasm07/POWER = 
18:59:44.683 RUL: MQTT#CONNECTED performs "Publish tele/tasm07/BLACKOUT {"MQTT":{"From":"2022-12-04T20:08:39","To":"2022-12-05T18:59:44"}}"
18:59:44.688 MQT: tele/tasm07/BLACKOUT = {"MQTT":{"From":"2022-12-04T20:08:39","To":"2022-12-05T18:59:44"}}
18:59:46.033 MQT: stat/tasm07/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Exception","Uptime":"1T20:25:59","StartupUTC":"2022-12-03T21:33:47","Sleep":50,"CfgHolder":4617,"BootCount":89,"BCResetTime":"2022-06-27T21:25:23","SaveCount":149,"SaveAddress":"F7000"}}
18:59:46.082 MQT: stat/tasm07/STATUS11 = {"StatusSTS":{"Time":"2022-12-05T18:59:46","Uptime":"1T20:25:59","UptimeSec":159959,"Heap":19,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":24,"MqttCount":31,"Wifi":{"AP":1,"SSId":"Oh Denneboom","BSSId":"******","Channel":1,"Mode":"11n","RSSI":74,"Signal":-63,"LinkCount":2,"Downtime":"0T00:00:05"}}}
18:59:46.132 MQT: stat/tasm07/STATUS10 = {"StatusSNS":{"Time":"2022-12-05T18:59:46","AM2301":{"Temperature":22.0,"Humidity":35.9,"DewPoint":6.2},"TempUnit":"C"}}
19:01:46.617 MQT: Attempting connection...
19:01:46.804 MQT: Connected
19:01:46.808 MQT: tele/tasm07/LWT = Online (retained)
19:01:46.810 MQT: cmnd/tasm07/POWER = 
19:01:46.832 RUL: MQTT#CONNECTED performs "Publish tele/tasm07/BLACKOUT {"MQTT":{"From":"2022-12-04T20:08:39","To":"2022-12-05T19:01:46"}}"
19:01:46.838 MQT: tele/tasm07/BLACKOUT = {"MQTT":{"From":"2022-12-04T20:08:39","To":"2022-12-05T19:01:46"}}
19:01:48.446 MQT: stat/tasm07/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Exception","Uptime":"1T20:28:01","StartupUTC":"2022-12-03T21:33:47","Sleep":50,"CfgHolder":4617,"BootCount":89,"BCResetTime":"2022-06-27T21:25:23","SaveCount":149,"SaveAddress":"F7000"}}
19:01:48.546 MQT: stat/tasm07/STATUS11 = {"StatusSTS":{"Time":"2022-12-05T19:01:48","Uptime":"1T20:28:01","UptimeSec":160081,"Heap":17,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":26,"MqttCount":32,"Wifi":{"AP":1,"SSId":"Oh Denneboom","BSSId":"******","Channel":1,"Mode":"11n","RSSI":72,"Signal":-64,"LinkCount":2,"Downtime":"0T00:00:05"}}}
19:01:48.597 MQT: stat/tasm07/STATUS10 = {"StatusSNS":{"Time":"2022-12-05T19:01:48","AM2301":{"Temperature":22.1,"Humidity":36.5,"DewPoint":6.5},"TempUnit":"C"}}
````

**Reply by barbudor** [reply]
there indeed no disconnect event

Interresting that the 2 connected events are exactly 2 minutes away from each other
Which is also the case for 3 of the events of your initial post
No idea for now what this could mean

**Comment by dheuts90** [comment]


**Reply by sciurius** [reply]
Maybe it is better to make this a separate discussion since it is not really related to the original question.

**Reply by dheuts90** [reply]
> Maybe it is better to make this a separate discussion since it is not really related to the original question.

Don’t you think it can be related? If not, sorry, I will make a new post.

**Reply by sciurius** [reply]
If .246 and .243 indicate version numbers it seems you just need to update your broker. It could be a glitch in .243 that has been fixed in .246.

**Reply by dheuts90** [reply]
Sorry for the confusion, but it are the last digits of the ip addresses of the raspberries running the mosquitto broker

**Reply by barbudor** [reply]
@dheuts90 this topic is not about reason of disconnection but handling of connection /disconnection events in rules
Please open a separated discussion


EDIT : I've seen your other topic. Please ignore my message. Thanks

**Comment by sciurius** [comment]
Attached a logfile (log level 5) that shows several connects and a lot less disconnects.
[tasm07.log](https://github.com/arendst/Tasmota/files/10170287/tasm07.log)

**Comment by arendst** [comment]
PubSubClient RC code definitions:
```
  // Possible values for state - PubSubClient.h
  Tasmota MQTT_DNS_DISCONNECTED       -5
  #define MQTT_CONNECTION_TIMEOUT     -4
  #define MQTT_CONNECTION_LOST        -3
  #define MQTT_CONNECT_FAILED         -2
  #define MQTT_DISCONNECTED           -1
  #define MQTT_CONNECTED               0
  #define MQTT_CONNECT_BAD_PROTOCOL    1
  #define MQTT_CONNECT_BAD_CLIENT_ID   2
  #define MQTT_CONNECT_UNAVAILABLE     3
  #define MQTT_CONNECT_BAD_CREDENTIALS 4
  #define MQTT_CONNECT_UNAUTHORIZED    5
```
Play with ``MqttTimeout`` if you expect connection issues.

**Comment by sciurius** [comment]
Increaded `MqttTimeout` from 4 to 10, no difference.
Note that the problem at hand is not so much disconnections, but that mqtt disconnections go unnoticed (i.e., do not fire rules).

**Comment by arendst** [comment]
Scanning through your log file I see that every MQTT disconnect with RC code als triggers the rule like:
```
Dec  5 19:23:42 tasm07 ESP-MQT: Connect failed to mqtt.squirrel.nl:1883, rc -2. Retry in 10 sec
Dec  5 19:23:42 tasm07 ESP-RUL: MQTT#DISCONNECTED performs "Var3 2022-12-05T19:23:42"
```
so I see no issue here....

**Comment by sciurius** [comment]
Yes, that's right. But there are also a lot of connects that do not have an associated disconnect.

**Reply by arendst** [reply]
Correct. What happens is that the loop checks for MQTT connections. If it fails it tries to reconnect. If that fails too it reports a disconnect. Otherwise it reports re-connect. So you'll see these messages popping up because the connection is lost and immediatly re-established.

Time to change your keep alive settings ``MqttKeepAlive`` and/or ``MqttWifiTimeout`` I guess.
